File: lib/src/bubble_options.dart
import 'enums/enums.dart';

/// [BubbleOptions] is a class that contains all the customizable options for the bubble passed to the [DashBubble.startBubble] method.
class BubbleOptions {
  /// The title of the service notification.
  final String? notificationTitle;

  /// The text of the service notification.
  final String? notificationText;

  /// The icon of the service notification.
  final String? notificationIcon;

  /// The icon of the bubble.
  final String? bubbleIcon;

  /// The icon of the close button.
  final String? closeIcon;

  /// The initial starting position of the bubble on the x axis.
  final double startLocationX;

  /// The initial starting position of the bubble on the y axis.
  final double startLocationY;

  /// The size of the bubble.
  final double bubbleSize;

  /// The opacity of the bubble and the close button.
  final double opacity;

  /// Whether to show the close button or not.
  final bool enableClose;

  final CloseBehavior closeBehavior;

  /// The distance between the bubble and the bottom edge of the screen to show the close button.
  final double distanceToClose;

  /// Whether to animate the bubble to the edge of the screen when it is dragged to the edge of the screen or not.
  final bool enableAnimateToEdge;

  /// Whether to show the bottom shadow behind the close button of the bubble or not.
  final bool enableBottomShadow;

  /// Whether to keep the bubble alive when the app is closed or not.
  final bool keepAliveWhenAppExit;

  BubbleOptions({
    this.notificationTitle,
    this.notificationText,
    this.notificationIcon,
    this.bubbleIcon,
    this.closeIcon,
    this.startLocationX = 0,
    this.startLocationY = 200,
    this.bubbleSize = 60,
    this.opacity = 1,
    this.enableClose = true,
    this.closeBehavior = CloseBehavior.following,
    this.distanceToClose = 100,
    this.enableAnimateToEdge = true,
    this.enableBottomShadow = true,
    this.keepAliveWhenAppExit = false,
  })  : assert(
          bubbleSize >= 0,
          'bubbleSize must be greater than or equal to 0',
        ),
        assert(
          opacity >= 0 && opacity <= 1,
          'opacity must be between 0 and 1',
        );

  Map<String, dynamic> toMap() {
    return {
      'notificationTitle': notificationTitle,
      'notificationText': notificationText,
      'notificationIcon': notificationIcon,
      'bubbleIcon': bubbleIcon,
      'closeIcon': closeIcon,
      'startLocationX': startLocationX,
      'startLocationY': startLocationY,
      'bubbleSize': bubbleSize,
      'opacity': opacity,
      'enableClose': enableClose,
      'closeBehavior': closeBehavior.index,
      'distanceToClose': distanceToClose,
      'enableAnimateToEdge': enableAnimateToEdge,
      'enableBottomShadow': enableBottomShadow,
      'keepAliveWhenAppExit': keepAliveWhenAppExit,
    };
  }
}

File: android/src/main/kotlin/dev/moaz/dash_bubble/src/BubbleOptions.kt
package dev.moaz.dash_bubble.src

import android.os.Parcelable
import io.flutter.plugin.common.MethodCall
import kotlinx.parcelize.Parcelize

/** BubbleOptions is the class that contains the options for the bubble.
 * @param bubbleIcon The icon of the bubble.
 * @param closeIcon The icon of the close button.
 * @param startLocationX The x coordinate of the bubble.
 * @param startLocationY The y coordinate of the bubble.
 * @param bubbleSize The size of the bubble.
 * @param opacity The opacity of the bubble.
 * @param enableClose Whether the bubble can be closed or not.
 * @param closeBehavior The behavior of the close button.
 * @param distanceToClose The distance between the bubble and the bottom edge of the screen to show the close button.
 * @param enableAnimateToEdge Whether the bubble can be animated to the edge or not.
 * @param enableBottomShadow Whether the bubble has a bottom shadow behind the close button or not.
 * @param keepAliveWhenAppExit Whether the bubble will be kept alive when the app is closed or not.
 */
@Parcelize
data class BubbleOptions(
    val bubbleIcon: String?,
    val closeIcon: String?,
    val startLocationX: Double?,
    val startLocationY: Double?,
    val bubbleSize: Double?,
    val opacity: Double?,
    val enableClose: Boolean?,
    val closeBehavior: Int?,
    val distanceToClose: Double?,
    val enableAnimateToEdge: Boolean?,
    val enableBottomShadow: Boolean?,
    val keepAliveWhenAppExit: Boolean?,
) : Parcelable {
    companion object {
        /**
         * Get the bubble options from the method call.
         * @param call The method call.
         * @return The bubble options.
         */
        fun fromMethodCall(call: MethodCall): BubbleOptions {
            return BubbleOptions(
                call.argument(Constants.BUBBLE_ICON),
                call.argument(Constants.CLOSE_ICON),
                call.argument(Constants.START_LOCATION_X),
                call.argument(Constants.START_LOCATION_Y),
                call.argument(Constants.BUBBLE_SIZE),
                call.argument(Constants.OPACITY),
                call.argument(Constants.ENABLE_CLOSE),
                call.argument(Constants.CLOSE_BEHAVIOR),
                call.argument(Constants.DISTANCE_TO_CLOSE),
                call.argument(Constants.ENABLE_ANIMATE_TO_EDGE),
                call.argument(Constants.ENABLE_BOTTOM_SHADOW),
                call.argument(Constants.KEEP_ALIVE_WHEN_APP_EXIT)
            )
        }
    }
}
